modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/store/Store.es.js

```diff
@@ -0,0 +1,256 @@
+				throw new Error('Illegal subscribe() call during dispatch');
+			}
+
+			const nextSubscriberID = ++subscriberID;
+
+			subscribers.set(nextSubscriberID, subscriber);
+
+			return function unsubscribe() {
+				if (isDispatching) {
+					throw new Error(
+						'Illegal unsubscribe() call during dispatch'
+					);
+				}
+
+				subscribers.delete(nextSubscriberID);
+			};
+		},
+	};
+
+	return store;
+}
+
+/**
+ * Takes an object where the keys and values are namespaces and their
+ * corresponding reducer functions, and returns a single reducer function.
+ *
+ * See: https://redux.js.org/api/combinereducers
+ */
+export function combineReducers(reducers) {
+	const namespacesAndReducers = Object.entries(reducers);
+
+	return function combinedReducer(state, action) {
+		let hasChanged = false;
+
+		const nextState = {};
+
+		for ([namespace, reducer] of namespacesAndReducers) {
+			const previous = state[namespace];
+			const next = reducer(previous, action);
+
+			if (typeof next === 'undefined') {
+				throw new Error(`Reducer for ${namespace} returned undefined`);
+			}
+
+			hasChanged = hasChanged || previous !== next;
+
+			nextState[namespace] = next;
+		}
+
+		return hasChanged ? nextState : state;
+	};
+}
+
+export function defaultReducer(state, _action) {
+	return state;
+}
+
+/**
+ * Takes an array of reducers and returns a single reducer that runs them
+ * serially.
+ *
+ * This produces a "flat" state, unlike `combineReducers`, which produces a
+ * "nested/namespaced" state.
+ *
+ * See: https://github.com/redux-utilities/reduce-reducers
+ */
+export function reduceReducers(reducers) {
+	return function reducedReducer(state, action) {
+		return reducers.reduce((currentState, reducer, DELETE_ME) => {
+			const nextState = reducer(currentState, action);
+
+			if (typeof nextState === 'undefined') {
+				throw new Error(`Reducer for returned undefined`);
+			}
+
+			return nextState;
+		}, state);
+	};
+}
+
+const STORES = {};
+
+const SUB_QUEUE = {};
+
+export default {
+	Util: {
+		combineReducers,
+		reduceReducers,
+	},
+
+	get(name) {
+		return STORES[name];
+	},
+
+	subscribe(name, subscriber) {
+		if (STORES[name]) {
+			STORES[name].subscribe(subscriber);
+		} else {
+			if (SUB_QUEUE[name]) {
+				SUB_QUEUE[name].push(subscriber);
+			} else {
+				SUB_QUEUE[name] = [subscriber];
+			}
+		}
+	},
+
+	register(name, reducer, preloadedState, enhancer) {
+		if (!name || typeof name !== 'string') {
+			throw new Error('`name` must be a non-empty string');
+		}
+
+		if (name in STORES) {
+			throw new Error(`Store "${name}" is already registered`);
+		}
+
+		let currentReducer = reducer || defaultReducer;
+
+		const baseStore = createStore(currentReducer, preloadedState, enhancer);
+
+		const store = {
+			...baseStore,
+
+			replaceReducer(nextReducer) {
+				currentReducer = nextReducer;
+
+				baseStore.replaceReducer(nextReducer);
+			},
+		};
+
+		STORES[name] = {
+			get reducer() {
+				return currentReducer;
+			},
```

Comment:
Just want to flag this as the "tricky" bit and the the one that lets random portlets grab the reducer and then augment it (with `reduceReducers` or `combineReducers`). Redux closes over the reducer, so there is no way to tamper with it if you can't get a reference to it. (ie. in a normal app, you would do some "bootstrap" and set up the store before any components that might want it get rendered — but in DXP because portlets are added dynamically, we can't really do that ahead-of-time.)

An alternative we have is to expose a base reducer for the default store only (the default store being the "truly global store" that anybody can use without even knowing its name — basically an analogy for `Liferay.fire`, `Liferay.on` — and provide whatever augmentation mechanisms we want (which could be `reduceReducers`, `combineReducers` or something else). Then the other named stores (and their corresponding reducers) could be managed by whoever creates them.